frontend/lib/src/WidgetStateManager.ts

```diff
@@ -709,6 +719,47 @@ export class WidgetStateManager {
       this.props.formsDataChanged(this.formsData)
     }
   }
+
+  /**
+   * Get the element state value for the given element ID and key, if it exists.
+   * This is a frontend-only state that is never sent to the server.
+   */
+  public getElementState(elementId: string, key: string): any {
+    return this.elementStates.get(elementId)?.get(key)
+  }
+
+  /**
+   * Sets the state of an element identified by its ID and its key.
+   * This is a frontend-only state that is never sent to the server.
+   * It can be used to store element state to restore the state
+   * of an element in situations where an element is removed and re-added.
+   *
+   * @param {string} elementId - The unique identifier of the element.
+   * @param {string} key - The key to set
+   * @param {any} value - The value to set for the element's state.
+   * @returns {void}
+   */
+  public setElementState(elementId: string, key: string, value: any): void {
+    if (!this.elementStates.has(elementId)) {
+      this.elementStates.set(elementId, new Map<string, any>())
+    }
+
+    this.elementStates.get(elementId)?.set(key, value)
```

Comment:
the `?` shouldn't be required here since the above logic makes sure that there is always a value for elementId, or?

Reply:
Typescript still wants the `?` as far as I can tell :( 
<img width="532" alt="Screenshot 2024-04-11 at 6 07 27 PM" src="https://github.com/streamlit/streamlit/assets/16749069/0df26c73-910e-466c-966e-03c5f171bad2">

Reply:
Hmm, I see. the type checker isn't smart enough :(

Reply:
Thanks! Added that.